data/build_redial.py

```diff
@@ -0,0 +1,156 @@
+import json
+from tqdm import tqdm
+import numpy as np
+import re
+import os
+from absl import app
+from absl import flags
+
+FLAGS = flags.FLAGS
+flags.DEFINE_string("data_dir", "./data/redial", "path to folder with redial data")
+flags.DEFINE_boolean("extra_redial_stats", False, "print extra summaries")
+
+RD_UNFORMATTED_FNAMES = {
+    "train": "rd-train-data.jsonl",
+    "validation": "rd-test-data.jsonl"
+}
+
+RD_FORMATTED_FNAMES = {
+    "train": "rd-train-formatted.jsonl",
+    "validation": "rd-test-formatted.jsonl"
+}
+
+def main(argv):
+    """Processes raw redial data in data_dir and saves the results in data folder."""
+    print("--Loading Redial Dataset--")
+    train = read_jsonl(os.path.join(FLAGS.data_dir, RD_UNFORMATTED_FNAMES["train"]))
+    test = read_jsonl(os.path.join(FLAGS.data_dir, RD_UNFORMATTED_FNAMES["validation"]))
+
+    print("--Replacing Movie IDs--")
+    for dialogue in tqdm(train):
+        replace_ids(dialogue)
+    for dialogue in tqdm(test):
+        replace_ids(dialogue)
+
+    print("-Formatting For Training--")
+    train_formatted = separate_responses(train)
+    test_formatted = separate_responses(test)
+
+    write_jsonl(os.path.join(FLAGS.data_dir, RD_FORMATTED_FNAMES["train"]), train_formatted)
+    write_jsonl(os.path.join(FLAGS.data_dir, RD_FORMATTED_FNAMES["validation"]), test_formatted)
+    
+    if FLAGS.extra_redial_stats:
+        length_summary(FLAGS.data_dir)
+
+def length_summary(data_dir):
+    """prints a five number summary of the lengths of redial input/outputs."""
+    print("--Loading Dataset For Summary--")
+    train_data = read_jsonl(os.path.join(data_dir, RD_FORMATTED_FNAMES["train"]))
+    test_data = read_jsonl(os.path.join(data_dir, RD_FORMATTED_FNAMES["validation"]))
+
+    def len_function(key):
+        return lambda x: len(x[key].split())
+
+    lengths = {"train_inputs" : list(map(len_function("conversation"), train_data)),
+                "train_targets" : list(map(len_function("response"), train_data)),
+                "test_inputs" : list(map(len_function("conversation"), test_data)),
+                "test_targets" : list(map(len_function("response"), test_data))}
+
+    for name, length_array in lengths.items():
+        print(name.upper() + ": ")
+        quartile_summary(length_array)
+
+#Helper Functions
+
+def read_jsonl(filename):
+    """Reads a jsonl file and returns as array of dicts."""
+    with open(filename, 'r') as json_file:
+        json_list = list(json_file)
+    data = []
+    for json_str in tqdm(json_list):
+        data.append(json.loads(json_str))
+    return data
+
+def write_jsonl(filename, arr):
+    """Writes array to jsonl."""
+    with open(filename, 'w') as f:
+        for line in arr:
+            f.write(json.dumps(line) + "\n")
+    
+def replace_ids(dialogue):
+    """Replaces movie ids in one redial dialogue with their corresponding movie titles. Each movie is surrounded by '@'
```

Comment:
I think it is a good idea to mention that the replacement is done "in-place", rather than returning a new dialog with the modified text.